feat(aws-elasticache): add ServerlessCache, User, and UserGroup L2 Constructs (#57)
open-constructs/aws-cdk-library: src/aws-elasticache/README.md

Constructs for the Amazon ElastiCache

# ElastiCache CDK Construct

This module has constructs for [Amazon ElastiCache](https://docs.aws.amazon.com/AmazonElastiCache/latest/dg/WhatIs.html).

> * The `User` and `UserGroup` construct facilitates the creation and management of user for the cache.

Reviewer: ```suggestion
* The `User` and `UserGroup` constructs facilitate the creation and management of user for the cache.
```

If you want to create a password authenticated user, use `PasswordUser` construct:

```ts
const user = PasswordUser(this, 'User', {
  // set user id
  userId: 'my-user-id',

  // set access string
  accessString: 'on ~* +@all',

  // set username
  userName: 'my-user-name',

  // set up to two passwords
  passwords: [
    cdk.SecretValue.unsafePlainText('adminUserPassword123'),
    cdk.SecretValue.unsafePlainText('anotherAdminUserPassword123'),
  ],
});
```

You can also create a no password required user by using `NoPasswordRequiredUser` construct:

```ts
const user = NoPasswordRequiredUser(this, 'User', {
  // set user id
  userId: 'my-user-id',

  // set access string
  accessString: 'on ~* +@all',

  // set username
  userName: 'my-user-name',
});
```

ElastiCache automatically configures a default user with user ID and username `default` and adds it to all user groups.
You can't modify or delete this user.

This user is intended for compatibility with the default behavior of previous Redis OSS versions and has an access string that permits it to call all commands and access all keys.

> To add proper access control to a cache, replace this default user with a new one that isn't enabled or uses a strong password.

Author: I think this refers to disabling the default user by setting `off -@all` in AccessString. 
This was quoted from [AWS documentation](https://docs.aws.amazon.com/AmazonElastiCache/latest/dg/Clusters.RBAC.html#rbac-using).

<img width="1066" alt="image" src="https://github.com/user-attachments/assets/62d8c548-7b83-4864-93f4-8a45bdcb950f" />
 

However, since it may be difficult to understand, I'll consider a revision.

To change the default user, create a new user with the username set to `default`. You can then swap it with the original default user.

For more information, see [Applying RBAC to a Cache for ElastiCache with Valkey or Redis OSS](https://docs.aws.amazon.com/AmazonElastiCache/latest/dg/Clusters.RBAC.html#rbac-using).

> If you want to create new default user, `userName` must be `default` and `userId` must not be `default` by using `NoPasswordRequiredUser` or `PasswordUser`:

Reviewer: Will CFn fail in the case of `userId` with `default` in `NoPasswordRequiredUser` or `PasswordUser`? Should we validate it?

> If you want to create new default user, `userName` must be `default` and `userId` must not be `default` by using `NoPasswordRequiredUser` or `PasswordUser`:

Author: Yes, the `default` user ID exists by default and can't be deleted.
<img width="1212" alt="image" src="https://github.com/user-attachments/assets/fed00978-8638-4a49-8211-72d79238216a" />

I've already added a validation to check whether the `userId` is `default` or not.

https://github.com/open-constructs/aws-cdk-library/blob/acfbe387ae58b6642b406f606a2c723f0eda45e5/src/aws-elasticache/user.ts#L159

> If you want to create new default user, `userName` must be `default` and `userId` must not be `default` by using `NoPasswordRequiredUser` or `PasswordUser`:

Reviewer: IAM missed it, that's OK!

Please only correct the sentence pointed out above.

> If you want to create new default user, `userName` must be `default` and `userId` must not be `default` by using `NoPasswordRequiredUser` or `PasswordUser`:

Author: I've corrected.

```ts
const newDefaultUser = NoPasswordRequiredUser(this, 'NewDefaultUser', {
  // default username must be 'default'
  userName: 'default',
  // new default user id must not be 'default'
  userId: 'new-default'
});
```

> > NOTE: You can't create a new default user using `IamUser` because an IAM-enabled user's username and user ID cannot be different.

Reviewer: Should we validate it?

Author: Since the validation throws an error when the `userId` is `default`, I think this requirement is already satisfied.

https://github.com/open-constructs/aws-cdk-library/blob/acfbe387ae58b6642b406f606a2c723f0eda45e5/src/aws-elasticache/user.ts#L159

### Add users to the user group

Next, create a user group by using `UserGroup` Construct and add users to the group:

```ts
declare const newDefaultUser: User;
declare const user: User;
declare const anotherUser: User;

const userGroup = new UserGroup(this, 'UserGroup', {
  // add users including default user
  users: [defaultUser, user],
});

// you can also add a user by using addUser method
userGroup.addUser(anotherUser);
```

### Assign user group

Finally, assign a user group to cache:

```ts
declare const userGroup: UserGroup;

const serverlessCache = new ServerlessCache(this, 'ServerlessCache', {
  engine: Engine.VALKEY,
  serverlessCacheName: 'my-serverless-cache',
  vpc,
  // assign User Group
  userGroup,
});

```

### Grant permissions to IAM-enabled users

> If you create IAM-enabled users, `“elasticache:Connect”` action must be allowed for user and cache.

Reviewer: ```suggestion
If you create IAM-enabled users, `“elasticache:Connect”` action must be allowed for the users and cache.
```

Reviewer: If the user is not an IAM user, do we need permissions to the cache?

If that story hasn't been mentioned anywhere yet, we might be able to add it (sorry if I missed it, I haven't seen the whole text yet).

Author: I have made the corrections and added that no password users and password authentication users do not require permissions.

For more information, see [Authenticating with IAM]((https://docs.aws.amazon.com/AmazonElastiCache/latest/dg/auth-iam.html)).

To grant permissions, you can use the `grantConnect` method in `IamUser` and `ServerlessCache` constructs:

```ts
declare const user: IamUser;
declare const serverlessCache: ServerlessCache;
declare const role: iam.Role;

// grant “elasticache:Connect” action permissions to Role
user.grantConnect(role);
serverlessCache.grantConnect(role);
```

### Import an existing user and user group

You can import an existing user and user group by using import methods:

```ts

> const importedIamUser = IamUser.fromUserAttributes(this, 'ImportedIamUser', 'my-iam-user-id');

Reviewer: If we should fix this README: 

```suggestion
const importedIamUser = IamUser.fromUserId(this, 'ImportedIamUser', 'my-iam-user-id');
```

const importedPasswordUser = PasswordUser.fromUserAttributes(stack, 'ImportedPasswordUser', {
  userId: 'my-password-user-id',
  userName: 'my-password-user-name',
});

const importedNoPasswordRequiredUser = NoPasswordRequiredUser.fromUserAttributes(stack, 'ImportedNoPasswordUser', {
  userId: 'my-no-password-user-id',
  userName: 'my-no-password-user-name',
});

> const importedUserGroup = UserGroup.fromUserGroupId(this, 'ImportedUser', 'my-user-group-id');

Reviewer: ```suggestion
const importedUserGroup = UserGroup.fromUserGroupId(this, 'ImportedUserGroup', 'my-user-group-id');
```

```

## Basic Usage for serverless cache

Setup required properties and create:

```ts
declare const vpc: ec2.Vpc;

const serverlessCache = new ServerlessCache(this, 'ServerlessCache', {
  engine: Engine.VALKEY,
  vpc,
  majorEngineVersion: MajorVersion.VER_8,
});
```

### Connecting to serverless cache

To control who can access the serverless cache by the security groups, use the `.connections` attribute.

> serverless cache have a default port `6379`.

Reviewer: ```suggestion
The serverless cache have a default port `6379`.
```

This example allows an EC2 instance to connect to the serverless cache:

```ts
declare const serverlessCache: ServerlessCache;
declare const instance: ec2.Instance;

// allow the EC2 instance to connect to serverless cache on default port 6379
serverlessCache.connections.allowDefaultPortFrom(instance);
```

The endpoint and the port to access your serverless cache will be available as the `.endpointAddress` and `.endpointPort` attributes:

```ts
declare const serverlessCache: ServerlessCache;

const endpointAddress = serverlessCache.endpointAddress;
const endpointPort = serverlessCache.endpointPort;
```

### Snapshots and restore

You can enable automatic backups for serverless cache.
When automatic backups are enabled, ElastiCache creates a backup of the cache on a daily basis.

Also you can set the backup window for any time when it's most convenient.
If you don't specify a backup window, ElastiCache assigns one automatically.

For more information, see [Scheduling automatic backups](https://docs.aws.amazon.com/AmazonElastiCache/latest/dg/backups-automatic.html).

> To enable automatic backups, set the `dailySnapshotTime` property to snapshot created time:

Reviewer: How about adding the description of the snapshotRetentionLimit?

Author: I've added a description about snapshotRetentionLimit.

```ts
const newDefaultUser = NoPasswordRequiredUser(this, 'NewDefaultUser', {
  // default username must be 'default'
  userName: 'default',
  // new default user id must not be 'default'
  userId: 'new-default'
});
```

> NOTE: You can't create a new default user using `IamUser` because an IAM-enabled user's username and user ID cannot be different.

### Add users to the user group

Next, create a user group by using `UserGroup` Construct and add users to the group:

```ts
declare const newDefaultUser: User;
declare const user: User;
declare const anotherUser: User;

const userGroup = new UserGroup(this, 'UserGroup', {
  // add users including default user
  users: [defaultUser, user],
});

// you can also add a user by using addUser method
userGroup.addUser(anotherUser);
```

### Assign user group

Finally, assign a user group to cache:

```ts
declare const userGroup: UserGroup;

const serverlessCache = new ServerlessCache(this, 'ServerlessCache', {
  engine: Engine.VALKEY,
  serverlessCacheName: 'my-serverless-cache',
  vpc,
  // assign User Group
  userGroup,
});

```

### Grant permissions to IAM-enabled users

> If you create IAM-enabled users, `“elasticache:Connect”` action must be allowed for the users and cache.

Reviewer: ```suggestion
If you create IAM-enabled users, `"elasticache:Connect"` action must be allowed for the users and cache.
```

> NOTE: You don't need grant permissions to no password required users or password authentication users.

For more information, see [Authenticating with IAM](https://docs.aws.amazon.com/AmazonElastiCache/latest/dg/auth-iam.html).

To grant permissions, you can use the `grantConnect` method in `IamUser` and `ServerlessCache` constructs:

```ts
declare const user: IamUser;
declare const serverlessCache: ServerlessCache;
declare const role: iam.Role;

> // grant “elasticache:Connect” action permissions to role

Reviewer: ```suggestion
// grant "elasticache:Connect" action permissions to role
```

user.grantConnect(role);
serverlessCache.grantConnect(role);
```

### Import an existing user and user group

You can import an existing user and user group by using import methods:

```ts
const importedIamUser = IamUser.fromUserId(this, 'ImportedIamUser', 'my-iam-user-id');

const importedPasswordUser = PasswordUser.fromUserAttributes(stack, 'ImportedPasswordUser', {
  userId: 'my-password-user-id',
  userName: 'my-password-user-name',
});

const importedNoPasswordRequiredUser = NoPasswordRequiredUser.fromUserAttributes(stack, 'ImportedNoPasswordUser', {
  userId: 'my-no-password-user-id',
  userName: 'my-no-password-user-name',
});

const importedUserGroup = UserGroup.fromUserGroupId(this, 'ImportedUserGroup', 'my-user-group-id');
```

## Basic Usage for serverless cache

Setup required properties and create:

```ts
declare const vpc: ec2.Vpc;

const serverlessCache = new ServerlessCache(this, 'ServerlessCache', {
  engine: Engine.VALKEY,
  vpc,
  majorEngineVersion: MajorVersion.VER_8,
});
```

> ### Connecting to serverless cache

Reviewer: How about talking about the grant method here?

Reviewer: Sorry, I missed that the IAM auth for the cache required with the IAM user.
